[precompile] Detect source code changes for save/load.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156432

Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

zhxchen17 committed:
[precompile] Detect source code changes for save/load.

Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Detect source code changes for save/load."


Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jamesjwu commented on torch/_dynamo/package.py:
```diff
@@ -184,12 +210,22 @@ def _initialize(self, fn: Any, dynamo: Optional[_DynamoCacheEntry] = None) -> No
+            for code in dynamo.inlined_sources:
+                m = importlib.import_module(code.module)
+                checksum = _hash_sourcelines(m, code.firstlineno, code.lastlineno)
+                if checksum != code.checksum:
```
I don't really get why we're RuntimeErroring; shouldn't we just use the compiler stance? If the compiler stance is not error on recompile, we should just ignore this package and recompile.

zhxchen17 replied:
If we treat this as cache, then it makes sense to ignore this. In an API where we call load() directly on a single model, imo we should always throw an error because load() shouldn't silently be a noop.

For load_cache_artifacts(), I don't have strong opinions whether to error or not (depending on how strong the guarantee is required to be).

I agree we should have a config somewhere governing the behavior here, not sure stance is a good option here because technically loading is not part of compilation today.

jamesjwu reviewed (commented):
I feel like inspect.getsourcelines is going to be slow on load. Triton does similar things and it ends up being a pretty large portion of compile time. Maybe this feature can be behind a config or option for each package?

zhxchen17 committed:
Update on "[precompile] Detect source code changes for save/load."


Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel reviewed (changes requested):
Test failures?

zhxchen17 committed:
Update on "[precompile] Detect source code changes for save/load."


Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Detect source code changes for save/load."


Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zhxchen17 commented:
@pytorchbot merge -i

zhxchen17 committed:
Update on "[precompile] Detect source code changes for save/load."


Go through all dynamo traced functions and compute checksum for them. While loading a precompilation back to memory, we will always check the checksum and refuse to load when
source code changes are detected.

Differential Revision: [D76987123](https://our.internmc.facebook.com/intern/diff/D76987123/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
